src/Main.elm

```diff
@@ -39,12 +52,15 @@ init _ =
 type alias Model =
     { rhymes : RhymeResult
     , phrases : PhraseResult
+    , wordInput : String
     , word : String
```

Comment:
Similar to the comment above, I wonder if it makes sense for the debouncer here to "own" both in-flight (typing) input as well as the last-finalized-because-user-stopped-typing-for-a-bit input, which would result in this attribute also being moved to the debouncer.

At that point, there's probably a union that can represent the different states here:

```elm
type Entering = Entering String
type Finished = Finished String

type DebouncerInput
    = NoInput
    | Typing Entering
    | Completed Finished
    | TypingAfterCompleted Finished Entering
```

To determine whether a request should be sent with a finished word, you could write:

```elm
finishedDebouncerInput : DebouncerInput -> Maybe String
finishedDebouncerInput input =
  case input of
    Completed (Finished v) -> Just v
    TypingAfterCompleted (Finished v) _ -> Just v
    _ -> Nothing
```

From below, `debouncerWaiting` could be written as:

```elm
debouncerWaiting : DebouncerInput -> Bool
debouncerWaiting input =
  case input of
    Finished _ -> False
    _ -> True
```

This is all pseudo-code and may not work 100%, but it would allow for removing `wordInput` and shifting `word` and the in-flight value to be owned by the debouncer itself.

Reply:
The ADS project has examples of going this route worth checking out. The `Dropdown` module owns all the behavior, its types look like this. There is some added complexity in this example (`Saveable` mainly) but just tune it out:

```elm
type alias Dropdown name =
    { label : String
    , state : DropdownState
    , name : name
    }


type alias OptionAtRest =
    Saveable (Maybe Option)


type DropdownState
    = Typing String OptionAtRest
    | Searching String OptionAtRest
    | DisplayingSearchResults String (ZipList Option) OptionAtRest
    | DisplayingNoSearchResults String OptionAtRest
    | AtRest OptionAtRest


{-| This represents one option in the dropdown. The `name` is what is displayed
to the user while the `id` is the value that will get saved.
-}
type alias Option =
    { id : Int
    , name : String
    }
```

Then in the `update` there's code for checking if after the debounce time is up, the value is still the same. If so it queues up the search and sets the state to `Searching`. Again, there's noise here but hopefully still useful to you in some shape:

```elm
update : Msg -> Config -> PageBody a b -> ( PageBody a b, Cmd Msg )
update msg config body =
    {- ... snip ... -}

      SearchDebounceComplete fieldName debouncedQuery ->
          searchIfApplicable config fieldName debouncedQuery body

searchIfApplicable : Config -> DropdownName -> String -> PageBody a b -> ( PageBody a b, Cmd Msg )
searchIfApplicable config fieldName query body =
    if fieldIsStillSearchingAndQueryMatches fieldName query body then
        body
            |> PageBody.modifyGlobalModal (changeDropdown fieldName (Dropdown.toSearching query))
            |> addCmd (fetchNamesLike config fieldName query)
```

We moved the repo on to the ADS github now but if you want to come by I'd be happy to look at code and chat further.